### PR TITLE
feat: register custom torch ops for in-repo kernels

### PR DESCRIPTION
**Pull request by winglian**
## Summary

Registers `torch.library` custom ops (namespace `axolotl::`) for the in-repo Triton kernel entry points that were previously invisible to the dispatcher. Two payoffs:

1. **torch.compile without graph breaks**: kernels become opaque ops with fake-tensor support instead of breaking the graph — verified with `fullgraph=True` compile tests (swiglu and the full `flash_d512` autograd.Function compile cleanly).
2. **Selective checkpointing compatibility** (#3786): policy-based SAC matches ops by name; `axolotl::flash_attn_d512_fwd` is picked up by the existing `attention` matcher as-is, and the dsv4/glm attention ops need only a one-line group extension (follow-up on that branch).

## What's registered (25 ops)

| Surface | Ops |
|:--|:--|
| `monkeypatch/attention/flash_attn_d512.py` | `flash_attn_d512_fwd/_bwd` |
| `kernels/swiglu.py`, `kernels/geglu.py` | `swiglu_fwd/_bwd`, `geglu_fwd/_bwd` (bwd = in-place mutation ops, `mutates_args`) |
| `kernels/rms_norm_gated.py` | `rms_norm_gated_fwd/_bwd` |
| `monkeypatch/trainer/utils.py` | `selective_log_softmax_fwd/_bwd`, `entropy_from_logits` |
| `core/trainers/ebft/kernels.py` | `ebft_fused_{log_softmax_gather,reinforce_loss,cosine_similarity,diversity_penalty}` |
| `integrations/kernels/libs/dsv4/` | `dsv4_{sliding,csa,csa_topk}_attn_fwd/_bwd` |
| `integrations/kernels/libs/glm_dsa/` | `glm_dsa_{mla_absorb,sparse_topk}_attn_fwd/_bwd` |

## Design

- `kernels/op_registry.py`: `register_kernel_op` decorator — `torch.library.custom_op` + warn-once fallback shim on registration failure (kernel keeps working, just unregistered).
- Op boundary = allocations + kernel launches only. Data-dependent host prep (varlen doc scans), dtype decisions, and non-schema args (callables, `torch.dtype` → crossed as bool) stay in the autograd.Functions, whose saved-tensor sets, ctx attrs, and public signatures are unchanged.
- Every op has a `register_fake` matching the real allocations exactly (incl. fp32 lse/delta buffers and post-cast grad dtypes).
- In-place backward kernels (swiglu/geglu) are declared with `mutates_args` and return `None`; the public wrappers return the mutated aliases *outside* the op (op outputs must never alias inputs). Auto-functionalization of this pattern under compile is verified by test.

## Not registered (deliberate)

scattermoe/sonicmoe autograd wrappers (their inner `scattermoe::*` ops are already registered; wrappers are aten-composed), comm wrappers (`_OverlapGather`, DeepEP), tiled-MLP orchestration Functions, and tiny elementwise helpers (`dora`, dsv4 `_RoPE`/`_GatedPool`/mhc mixers).

## Verification

- 474 tests green across all touched surfaces (existing suites now exercising the op-routed paths), incl. new coverage where none existed (dsv4 attention parity vs dense fp32 sink-softmax reference, glm_dsa sparse-topk parity).
- `tests/kernels/test_op_registry.py`: op existence, `TorchDispatchMode` visibility, fake-tensor shapes, `torch.library.opcheck` (schema/aliasing/mutation/fake in one shot), `fullgraph=True` compile parity (incl. the mutating-bwd functionalization case), and the registration-failure fallback.
- Adversarial kernel-correctness review of the full diff; its one substantive finding (mutate-and-read-back under compile) was empirically refuted and is now a pinned regression test; its fake-stride-fidelity nit is fixed.

## Summary by CodeRabbit

* **New Features**
  * Added broader registration support for custom GPU kernels, improving integration with PyTorch dispatch and compile paths.
  * Added fake-tensor support for several attention and fused MLP operations, enabling shape inference without real execution.

* **Bug Fixes**
  * Improved compatibility for multiple attention, normalization, and activation kernels across forward and backward passes.
  * Standardized several kernel entry points for more reliable runtime behavior.

* **Tests**
  * Added CUDA integration, parity, dispatch, fake-tensor, compile, and registration coverage for the new kernel ops.

## E2E compile smoke (manual verification — NOT in CI; committed tests use only small synthetic tensors)

- **Gemma-4-E2B (MoE) + kernels plugin + LoRA + `torch_compile: true`**: trains cleanly; dynamo captures 3988 ops in 5 graphs; all 8 graph breaks are PEFT/HF-side (none in kernel code).
- **Gemma-4-26B-A4B + `gemma4_hybrid_attn_impl` + sample packing** (the d512 kernel's production path): the `axolotl::flash_attn_d512` op executes in-model under compile (Triton autotune observed in the training step; 435 ops / 37 graphs). The only axolotl-attributed graph breaks are two pre-existing data-dependent lines in the routing wrapper (`large_head.py:52/:63` multidoc detection), not the registered ops — follow-up candidate to hoist out of the traced region.

### CODE REVIEW

File: src/axolotl/kernels/rms_norm_gated.py [hunk 1]
```diff
@@ -267,9 +280,45 @@ def rms_norm_gated_backward(dY, X, G, W, RSTD, offset, BLOCK_SIZE, num_warps):
...
+    assert X.shape[1] == W.shape[0], (
+        f"Incompatible hidden size: X.shape[1]={X.shape[1]} vs W.shape[0]={W.shape[0]}"
+    )
+    assert X.shape == G.shape, (
+        f"X and G must have same shape, got {X.shape} and {G.shape}"
+    )
+
+    Y, RSTD = _rms_norm_gated_fwd_op(X, G, W, eps, offset, BLOCK_SIZE, num_warps)
+    return Y.view(*shape), X, G, RSTD, BLOCK_SIZE, num_warps
+
```
[the remark this answers is not in the document]
Comment (winglian) [reply src/axolotl/kernels/rms_norm_gated.py:308]: Fixed in 5ed05456f — dropped the unused unpack (`calculate_settings(X.shape[-1])`).

**Comment by winglian** [comment]
Addressed the CodeRabbit feedback in 5ed05456f:

- **rms_norm_gated**: removed the unused `n_rows` unpack (RUF059).
- **op_registry**: warn-once is now actually implemented — first registration failure logs at WARNING, subsequent ones at DEBUG (matches the PR description).
- **swiglu/geglu/mla_absorb comments**: condensed to one line per the repo comment guideline.
- **Skipped**: the E731 lambda-assignment suggestion in `attention_gather.py` — E731 is deliberately ignored repo-wide (`pyproject.toml:202`, "idiomatic for the kernel recipe/heuristic closures") and the flagged lines carry explicit `# noqa: E731` matching the surrounding kernel code style.
